File: arrays/max-distance.cpp
×
int Solution::maximumGap(const vector<int> &A) {
    vector<pair<int,int>> v;
    int n = A.size();
    for(int i=0;i<n;i++) v.push_back({A[i],i});
    
    sort(v.begin(),v.end());
    
    int gmax=v[0].second, gmin=v[0].second;
    int lmax=gmax, lmin=gmin;
    
    for(int i=1; i<n; i++){
        lmax = max(lmax, v[i].second);
        if(v[i].second<lmin){
            if(gmax-gmin < lmax-lmin){
                gmax = lmax;
                gmin = lmin;
            }
            lmin = v[i].second;
            lmax = v[i].second;
        }
    }
    
    if(gmax-gmin < lmax-lmin){
        gmax = lmax;
        gmin = lmin;
    }
    // if(gmax==gmin) return -1;
    return gmax-gmin;
}

File: arrays/max-distance.py
class Solution:
    def maximumGap(self, A):
        A = list(zip(A, range(len(A))))
        A.sort()
        l = sys.maxsize
        ans = 0

        for x in A:
            l = min(l, x[1])
            ans = max(ans, x[1]-l)

        return ans

#Alternate

class Solution:
    def maximumGap(self, A):
        maxA = [0]*len(A)
        maxA[len(A)-1] = A[len(A)-1]
        for i in range(len(A)-2, -1, -1): maxA[i] = max(A[i], maxA[i+1])

        i = 0; j = 0; ans = 0
        while i < len(A) and j < len(A):
            if A[i] <= maxA[j]:
                ans = max(ans, j-i)
                j+=1
            else: i+=1

        return ans;
